packages/mds-ingest-service/repository/index.ts

```diff
@@ -472,20 +472,30 @@ class IngestReadWriteRepository extends ReadWriteRepository {
 
   public getLatestTelemetryForDevices = async (device_ids: UUID[]): Promise<TelemetryDomainModel[]> => {
     try {
-      const connection = await this.connect('ro')
-      const entities = await connection
-        .createQueryBuilder(TelemetryEntity, 'telemetry')
-        .innerJoin(
-          subquery =>
-            subquery
-              .select('id, RANK() OVER (PARTITION BY device_id ORDER BY timestamp DESC) AS rownum')
-              .where('device_id = ANY(:device_ids)', { device_ids })
-              .from(TelemetryEntity, 't'),
-          'last_device_telemetry',
-          'last_device_telemetry.id = telemetry.id AND last_device_telemetry.rownum = 1'
-        )
-        .getMany()
-      return entities.map(TelemetryEntityToDomain.mapper())
+      /**
+       * https://github.com/typeorm/typeorm/blob/master/docs/transactions.md#using-queryrunner-to-create-and-control-state-of-single-database-connection
+       */
+      const runner = await (await this.connect('ro')).createQueryRunner()
+
+      await runner.startTransaction()
+      try {
+        /* Docs: set_config sets the parameter setting_name to new_value. If is_local is true, the new value will only apply to the current transaction
+           Sets enable_bitmapscan off for this transaction/query
+        */
+        await runner.query(`select set_config('enable_bitmapscan','off', true)`)
+
+        const entities = await runner.manager
+          .createQueryBuilder(TelemetryEntity, 'telemetry')
+          .distinctOn(['device_id'])
+          .orderBy('device_id', 'DESC')
+          .addOrderBy('timestamp', 'DESC')
+          .where('device_id = ANY (:device_ids)', { device_ids })
+          .getMany()
+        return entities.map(TelemetryEntityToDomain.mapper())
+      } finally {
```

Comment:
I generally prefer `finally` to appear after the `catch` since it’s called last. Does the catch need to rollback explicitly or is that automatic when the transaction goes out of scope without being committed?

Reply:
I think in this case, there's actually no explicit `catch` around this _particular_ block, it defers all catches up to the top-level `try/catch` that's wrapping this whole function.